### ISSUE TITLE
How to upload result to coveralls?

### ISSUE THREAD

**Issue by fschutt** [SimonKagstrom/kcov#269]
Right now I'm trying to switch my project to coveralls for code coverage. However, I don't get how I am supposed to upload the results to coveralls or what keys to set - I know this isn't related to kcov directly, however, I've set the `coveralls-id` - but nothing shows up on coveralls. Their documentation doesn't explain anything, sadly. What else do I need to set up to make coveralls work with kcov?

Eventually I've managed to get it somewhat working: https://travis-ci.org/maps4print/azul/jobs/433814003#L936 - the config for this is [here](https://github.com/maps4print/azul/blob/def068344dd5612f7360c19959085a943f22a6b1/.travis.yml#L41-L49), I directly copied it from the [glium configuration](https://github.com/glium/glium/blob/cda2cc0229f430baa72b49a3e407c153221b3c8d/.travis.yml#L59-L67) - what I don't get is why it [works for glium](https://coveralls.io/github/tomaka/glium), but not for me. It's the exact same command, but for some reason [I don't get any code coverage](https://coveralls.io/github/maps4print/azul):

```yaml
after_success:
- |
  [ $TRAVIS_BRANCH = master ] &&
  [ $TRAVIS_PULL_REQUEST = false ] &&
  wget https://github.com/SimonKagstrom/kcov/archive/master.zip &&
  unzip master.zip && mv kcov-master kcov && mkdir kcov/build && cd kcov/build &&
  cmake .. && make && make install DESTDIR=../built && cd ../.. &&
  for file in ./target/debug/*; do ./kcov/built/usr/local/bin/kcov --verify --exclude-pattern=/.cargo,$PWD/target/debug/build,$PWD/tests ./target/kcov ${file}; done &&
  ./kcov/built/usr/local/bin/kcov --coveralls-id=${TRAVIS_JOB_ID} --merge ./target/kcov-merge ./target/kcov
```

- Any idea, just from the configuration, why I don't get any results?
- What does `--coveralls-id` actually do?
- Does kcov upload the results? If not, why or how does it need the `--coveralls-id`? And how do I upload them?
- coveralls just shows me a "repo token", do I need it for kcov? I set it as a secret `COVERALLS_REPO_TOKEN` in travis, but that does absolutely nothing.
- coveralls says I need a `.coveralls.yml`, but glium has no `.coveralls.yml`, yet it still works?
- Do I need to set some "secure" tokens, and if yes, where?
- Does kcov look for any tokens? Which ones? Where do I get these tokens from?
- Why does kcov say "X invalid breakpoints skipped"?
- Where are my actual code coverage files (I assume `./target/kcov`)?

For codecov, I had this configuration:

```yaml
- |
wget https://github.com/SimonKagstrom/kcov/archive/v34.tar.gz &&
tar xzf v34.tar.gz &&
cd kcov-34 &&
mkdir build &&
cd build &&
cmake .. &&
make &&
sudo make install &&
cd ../.. &&
rm -rf kcov-34 &&
find target/debug -maxdepth 1 -name 'azul-*' -type f | while read file; do
  [ -x $file ] || continue;
  mkdir -p "target/cov/$(basename $file)";
  kcov --exclude-pattern=/.cargo,/usr/lib  --include-path="$(pwd)" --verify "target/cov/$(basename $file)" "$file";
done &&
bash <(curl -s https://codecov.io/bash) &&
echo "Uploaded code coverage"
```

... which has a clear `bash <(curl -s https://codecov.io/bash)` to upload the results... but for coveralls? Since kcov knows about coveralls, maybe the uploading part is built-in? In short, I'm more or less just fumbling in the dark trying to get it working. Thanks for any hints.

**Comment by SimonKagstrom** [comment]
`--coveralls-id=` sets the repository token, see https://docs.coveralls.io/api-introduction. On travis-ci builds, it uses the travis build ID as you can see above, but you can also use it manually from the command line, when you would use the repository token directly.

I can't immediately see why your build fail, though. Perhaps one thing could be the curly braces around `${TRAVIS_JOB_ID}.` Looking at e.g., https://gist.github.com/jacklund/0ce13d9869c7371d763c05f9388c87fb, they only use `$TRAVIS_JOB_ID` from after_success.

As to your other questions:

- Yes, kcov uploads to coveralls. Make sure libcurl4-openssl-dev is installed (which it appears to be from your logs)
- The secret repo token can be used from your own computer, but should not be used from Travis builds.
- kcov itself doesn't have a `.coveralls.yml`, so that should be fine :-)
- As for the invalid breakpoints, see for example #212 where this is discussed
- Your script runs all binaries and places the output in `./target-kcov`. Then it merges that output into `./kcov-merge` and uploads that to coveralls.

I personally don't upload to coveralls via `--merge` (because I don't need to), but I don't see why it shouldn't work. If gilium does this, I believe it should work.

In short, try removing braces first :-)

**Comment by fschutt** [comment]
I finally got it to at least build, but it seems that the problem wasn't the curly braces, it doesn't matter if you have them or not. What actually matters is that you run the kcov command with `$(find target/debug -maxdepth 1 -executable -name my_project_name-\*)` instead of a for loop with `--merge`. `find` will skip any non-executable files (which is good, so you don't get any errors) and select the executables by name. It seems that kcov only uploads if there haven't been any errors (?).

I ran it with the command from jacklund and it finally worked, i.e. instead of:
```
for file in ./target/debug/*; do 
    ./kcov \
     --verify 
     --exclude-pattern=/.cargo,$PWD/target/debug/build,$PWD/tests ./target/kcov ${file}; 
done
```

I did:

```
  kcov \
--verify 
--coveralls-id=$TRAVIS_JOB_ID \
--exclude-pattern=/.cargo target/kcov $(find target/debug -maxdepth 1 -executable -name azul-\*)
```

It seems that in order to work correctly, kcov has to run without any errors.

The only problem I have now is that it skips all tests and therefore gives me a 0% code coverage:

```
Install the project...
-- Install configuration: "Release"
-- Installing: /usr/local/share/doc/kcov/ChangeLog
-- Installing: /usr/local/share/doc/kcov/COPYING
-- Installing: /usr/local/share/doc/kcov/COPYING.externals
-- Installing: /usr/local/bin/kcov
-- Installing: /usr/local/bin/kcov-system-daemon
-- Installing: /usr/local/share/man/man1/kcov.1

running 0 tests
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 64 filtered out

kcov: 1776 invalid breakpoints skipped in /home/travis/build/maps4print/azul/target/debug/azul-fc50e6937e4ed0e1

Done. Your build exited with 0.
```

So it finds the testing executable, but then it just... skips it. I'm not sure how that happened, [log](https://travis-ci.org/maps4print/azul/jobs/434181295#L1349) for reference.

**Comment by SimonKagstrom** [comment]
Ah!

Well, it's not kcov but the shell that does that. You had

``` for file in ./target/debug/*; do [...]; done &&
  ./kcov/built/usr/local/bin/kcov [...] ./target/kcov-merge ./target/kcov```

and if one of the kcov executions in the loop exits with a non-zero code, the merge command will never execute because of the &&. Perhaps better would be to have a ; instead of the &&, or do as you did.


As for the filtered tests, that looks like an issue with the testsuite, since  this isn't output from kcov. I'm not sure why that happens. An alternative could be to try replacing the && in the original command line with a ;, so that coverage is uploaded also on failures.

**Comment by fschutt** [comment]
Two weeks ago, these issues "fixed" themselves - I didn't change anything about my travis.yml, however, someone made a pull request to this repository. It seems that coveralls then recognized: "oh, I should actually process these results and display them". I did not change the script otherwise. So this does seem to be an issue on coveralls side, to solve it, just make and accept an empty PR.

**Comment by SimonKagstrom** [comment]
Ok! Great to hear, although a bit surprising :-)
